### PR TITLE
added file checksum in annotation metadata

### PR DESCRIPTION
**Pull request by cage2**
Hello!

With these commits i tried to address #6. I have added am md5 hash to each annotated file. When  annotate minor mode is loaded a check between the hash stored in the metadata (`.emacs.d/annotations`) and the current buffer is performed. If the two checksum are not equals a warning text is displayed.
When a file is saved an md5 sum is calculated for current buffer and stored in the metadata file.

Also added a 'blacklist' of major modes that prevents annotate to be loaded.

Hope i understood #6 correctly, though! :-)

Please consider merging.

bye!
C.

### CODE REVIEW

File: annotate.el
Comment (bastibe) [inline annotate.el]: ...the annotations *were* saved...
...to warn *the* user that *the* file has been modified and *annotation positions* could...

File: annotate.el
Comment (bastibe) [inline annotate.el]: Why did you add this constraint?

File: annotate.el
Comment (bastibe) [inline annotate.el:129]: This was discussed in a different issue as well. Do you know if the other issue made any progress on this?

File: annotate.el
Comment (bastibe) [inline annotate.el:143]: This function should not be called `-hook`, as that is commonly used for variables that may *contain* functions, not the functions themselves. How about `annotate-initialize-maybe`?

File: annotate.el
Comment (bastibe) [inline annotate.el:267]: Thank you for adding documentation!

File: annotate.el
Comment (bastibe) [inline annotate.el:127]: ...loading of *annotate-mode*...

File: annotate.el
Comment (bastibe) [inline annotate.el]: Does annotate.el already require `cl`? If not, could this be implemented without `cl`?

Comment (bastibe) [reply annotate.el]: Please add a docstring.

File: annotate.el
Comment (bastibe) [inline annotate.el:157]: Please add a docstring.

File: annotate.el
Comment (bastibe) [inline annotate.el:177]: Please add a docstring.

File: annotate.el
Comment (bastibe) [inline annotate.el:285]: ...(if *it* makes sense) *end*...

File: annotate.el
Comment (bastibe) [inline annotate.el:722]: Please add docstring.

File: annotate.el
Comment (bastibe) [inline annotate.el:728]: Please add docstring.

**Review by bastibe — CHANGES_REQUESTED** [review]
Thank you so much for this pull request! This is exactly what I had in mind in #6.

Before merging, I would request a few small changes. In addition to the changes mentioned in the review comments, please go through all changed docstrings and delete any double-spaces.

Again, thank you very much. I see you implemented it without breaking backwards compatibility, too! Great job!

**Comment by cage2** [comment]
Hi Bastibe!

> Thank you so much for this pull request! This is exactly what I had in mind in #6.

Oh, that's great! :)
 
> Before merging, I would request a few small changes. In addition to the changes mentioned in the review comments, please go through all changed docstrings and delete any double-spaces.

Sure! No problem! :)

> 
> Again, thank you very much. I see you implemented it without breaking backwards compatibility, too! Great job!

To be honest i /tried/ not breaking the compatibility but i sadly discovered i failed to accomplish that. :(
Anyway i just pushed a quick hack that should fix that! Not the most elegant code i have seen but better that keeping a bug, i guess ;-)

Bye! 
C.

File: annotate.el
Comment (cage2) [reply annotate.el]: Thank you for ~~fix~~ fixing my bad english!

File: annotate.el
Comment (cage2) [reply annotate.el]: Because, i think, if the file is very large the calculation could take a lot of time so maybe an user could just skips this step and save time even if they know that annotation could be "out of sync".

File: annotate.el
Comment (cage2) [reply annotate.el:129]: I still have problem with global mode but i think adding this variable was a good idea anyway because the incompatibility with org-mode is a known problem and this setting could prevent a bit of frustration for an user that do not read documentation with attention (like me for example)! ;-D

File: annotate.el
Comment (cage2) [reply annotate.el:143]: nice!

File: annotate.el
Comment (cage2) [reply annotate.el:267]: i hate writing documentation (like many others :-)) but i hate more code that is not documented!  ;-D

File: annotate.el
Comment (cage2) [reply annotate.el]: IIRC cl-defun is needed because of the default optional argument

https://www.gnu.org/software/emacs/manual/html_mono/cl.html#Argument-Lists

So maybe the right question is: "is an optional argument needed at all?" :)

File: annotate.el
Comment (bastibe) [reply annotate.el]: I would suspect that md5 will never take remotely as long as Emacs will take to load the file in the first place. After all, md5 is routinely used to checksum gigabytes worth of data.

Did you have an actual example where big files were a problem? Otherwise, I'd say we go without the added complexity of this constraint (for now). We can still add them later if it turns out to be a problem for users.

If that's ok with you, of course.

File: annotate.el
Comment (bastibe) [reply annotate.el:129]: OK, cool.

File: annotate.el
Comment (bastibe) [reply annotate.el]: Silly me. I didn't even realize that annotate.el *already* requires `cl`. So this line of inquiry was pointless. Sorry about that.

**Comment by bastibe** [comment]
As far as I can tell, the only open issue remains whether to remove the size restriction or not. I'll leave it up to you to keep it in the code or not.

Let me know when you think this is ready to merge!

File: annotate.el
Comment (cage2) [reply annotate.el]: Hello!

i am fine with removing the constraint, i have made a (very naive) test and -for a 5 years old computer- a ~200Mb file takes about 0.4s to calculate the hash and a couple of second (very approximate estimation!) to open in emacs, i am OK to remove the size check as i think is unusual to have more than a couple of files so big and i bet modern system got better calculating hash.

Maybe we could inform (in the README, for example) the user that such calculation occurs because, for example, i use desktop mode that restores all the buffers opened in a previous emacs session and is not unlikely (at least for me :)) to get restored more than 20 buffer or so. Anyway i just pushed the fix with the check removed and we can save the README update for another PR, if needed.

Bye!
C.

Comment (bastibe) [reply annotate.el]: Alright!

**Comment by bastibe** [comment]
Thank you very much for your contribution, and our enjoyable conversation.

**Comment by cage2** [comment]
My pleasure! :)

Bye!
C.

**Pull request merged**
